### PR TITLE
Plutip core

### PR DESCRIPTION
**Pull request by zmrocze**
Strip down plutip to plutip-core

This change leaves plutip as a tool for spawning local clusters,
removing all the testing functionalities for testing contracts.

 - remove functionality to run test contracts
 - sync deps to cardano-node-1.35.4
 - remove BPI and plutus-apps from deps
 - remove setup for BPI from cluster setup
 - refactor cluster setup
 - new: withCluster (previously withPlutusInterface), withFundedCluster
 - change module structure
 
 ---------------------------
  - [x] Update Changelog
  - [x] Format
  - [x] Update docs
  - [ ] Refactor nix 
----------------------------

### CODE REVIEW

**Comment by zmrocze** [comment]
Specifically changes to `plutip-server`:
 - fix reporting `NegativeLovelace` (previously this failed too early in json parsing)
 - remove handling of `PlutipServerError` as it was never thrown
 - TODO ex units
 
Good improvement would be to report in `ClusterStartupFailureReason` a case when wallet funding fails due to requested utxo distribution not satisfying the minimum Ada requirement. When this happens the cluster raises [this exception](https://github.com/mlabs-haskell/plutip/blob/b683b3d10f3903158b71044a3330038f0874fcc7/src/Plutip/Launch/Cluster.hs#L430).

File: src/Plutip/Launch/Cluster.hs [hunk 1]
```diff
@@ -2436,6 +1969,178 @@ instance HasSeverityAnnotation ClusterLog where
...
+    --     traceCounters trace cts
+    --     threadDelay 30_000_000   -- 30 seconds
+    --   pure ()
+    --  where
+    --    traceCounters :: forall m a. MonadIO m => Trace m a -> [Counter] -> m ()
+    --    traceCounters _tr [] = return ()
+    --    traceCounters tr (c@(Counter _ct cn cv) : cs) = do
+    --      mle <- mkLOMeta Notice Confidential
+    --      traceNamedObject tr (mle, LogValue (nameCounter c <> "." <> cn) cv)
+    --      traceCounters tr cs
```
Comment (klntsky) [inline src/Plutip/Launch/Cluster.hs:2067]: What value does this commented-out code provide?

File: src/Plutip/DistributeFunds.hs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import Cardano.Api (Lovelace (Lovelace))
+import Plutip.Keys (KeyPair, mainnetAddress)
+
+-- | Fund key's enterprise address with a transaction spending faucet funds.
+fundKey :: ClusterEnv -> KeyPair -> [Lovelace] -> IO ()
+fundKey cEnv keys funds = do
+  let fundAddress = mainnetAddress keys
+      toAmt = Coin . fromIntegral
+  sendFaucetFundsTo
+    nullTracer -- todo: fix tracer to be not `nullTracer`
```
Comment (klntsky) [inline src/Plutip/DistributeFunds.hs:21]: Why this TODO? What's a better option than nullTracer?

File: src/Plutip/Cluster.hs [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+data ClusterStatus (a :: Type)
+  = ClusterStarting
+  | ClusterStarted a
+  | ClusterClosing
+  | ClusterClosed
+
+data StopClusterRef = forall a. StopClusterRef (TVar (ClusterStatus a))
+
+isPrefixOf :: (Eq a) => [a] -> [a] -> Bool
```
Comment (klntsky) [inline src/Plutip/Cluster.hs:294]: It exists in `Data.List`

File: src/Plutip/Cluster.hs [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+withFundedCluster :: PlutipConfig -> [[Lovelace]] -> (ClusterEnv -> [KeyPair] -> IO a) -> IO a
+withFundedCluster conf distribution action = withCluster conf $ \cenv -> do
+  keys <- for distribution (const genKeyPair)
+  zipWithM_ (fundKey cenv) keys distribution
+  for_ keys (fmap dieOnError . saveKeyPair (keysDir cenv))
+  awaitUtxosNumber
+    cenv
+    (cardanoMainnetAddress <$> keys)
+    (length $ concat distribution)
+    >>= dieOnError
```
Comment (klntsky) [inline src/Plutip/Cluster.hs:92]: We should not terminate, we need to throw.

**Review by klntsky — CHANGES_REQUESTED** [review]
Thanks Karol! We will handle these comments by ourselves.

**Pull request merged**